pkg/spanconfig/spanconfigmanager/manager.go

```diff
@@ -0,0 +1,147 @@
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package spanconfigmanager
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/cockroachdb/cockroach/pkg/jobs"
+	"github.com/cockroachdb/cockroach/pkg/jobs/jobspb"
+	"github.com/cockroachdb/cockroach/pkg/kv"
+	"github.com/cockroachdb/cockroach/pkg/security"
+	"github.com/cockroachdb/cockroach/pkg/spanconfig"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/sessiondata"
+	"github.com/cockroachdb/cockroach/pkg/sql/sqlutil"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/stop"
+)
+
+// Manager is responsible for ensuring that one (and only one) span config
+// reconciliation job is ever running. It also provides the job with all the
+// components it needs to actually perform reconciliation.
+type Manager struct {
+	db      *kv.DB
+	jr      *jobs.Registry
+	ie      sqlutil.InternalExecutor
+	knobs   *spanconfig.TestingKnobs
+	stopper *stop.Stopper
+}
+
+var _ spanconfig.ReconciliationDependencies = &Manager{}
+
+// New constructs a new Manager.
+func New(
+	db *kv.DB,
+	jr *jobs.Registry,
+	ie sqlutil.InternalExecutor,
+	stopper *stop.Stopper,
+	knobs *spanconfig.TestingKnobs,
+) *Manager {
+	if knobs == nil {
+		knobs = &spanconfig.TestingKnobs{}
+	}
+	return &Manager{
+		db:      db,
+		jr:      jr,
+		ie:      ie,
+		knobs:   knobs,
+		stopper: stopper,
+	}
+}
+
+// StartJobIfNoneExists will create and start the span config reconciliation job
+// iff it hadn't been created already. Returns a boolean indicating if the job
+// was started.
+func (m *Manager) StartJobIfNoneExists(ctx context.Context) bool {
+	if m.knobs.ManagerDisableJobCreation {
+		return false
+	}
+	started, err := m.createAndStartJobIfNoneExists(ctx)
+	if err != nil {
+		log.Errorf(ctx, "error starting auto span config reconciliation job: %v", err)
+	}
+	if started {
+		log.Infof(ctx, "started auto span config reconciliation job")
+	}
+	return started
+}
+
+// createAndStartJobIfNoneExists creates span config reconciliation job iff it
+// hasn't been created already and notifies the jobs registry to adopt it.
+// Returns a  boolean indicating if the job was created.
+func (m *Manager) createAndStartJobIfNoneExists(ctx context.Context) (bool, error) {
+	record := jobs.Record{
+		JobID:         m.jr.MakeJobID(),
+		Description:   "reconciling span configurations",
+		Username:      security.RootUserName(),
+		Details:       jobspb.AutoSpanConfigReconciliationDetails{},
+		Progress:      jobspb.AutoSpanConfigReconciliationProgress{},
+		NonCancelable: true,
+	}
+
+	var job *jobs.Job
+	if err := m.db.Txn(ctx, func(ctx context.Context, txn *kv.Txn) error {
+		exists, err := m.checkIfReconciliationJobExists(ctx, txn)
```

Comment:
@Azhng has a thing for this coming in https://github.com/cockroachdb/cockroach/pull/68434/, see `RunningJobExists`.